### PR TITLE
[InstSimplify][InstCombine][ConstantFold] Move vector div/rem by zero fold to InstCombine

### LINKED ISSUE #114191: [InstCombine] miscompilation of `or` of vector
Alive2 report: https://alive2.llvm.org/ce/z/xEbYKg

```llvm
----------------------------------------
define <2 x i32> @or_sext_to_sel_constant_vec.2(<2 x i1> %y) {
#0:
  %sext = sext <2 x i1> %y to <2 x i32>
  %#1 = or <2 x i32> %sext, { 0, 1 }
  %#2 = sdiv <2 x i32> { 42, 4294967289 }, %#1
  %#3 = udiv exact <2 x i32> %#2, %#2
  %#4 = or <2 x i32> %#2, %#2
  ret <2 x i32> %#4
}
=>
define <2 x i32> @or_sext_to_sel_constant_vec.2(<2 x i1> %y) {
#0:
  ret <2 x i32> { 4294967254, 7 }
}
Transformation doesn't verify!

ERROR: Value mismatch

Example:
<2 x i1> %y = < #x1 (1), #x0 (0) >

Source:
<2 x i32> %sext = < #xffffffff (4294967295, -1), #x00000000 (0) >
<2 x i32> %#1 = < #xffffffff (4294967295, -1), #x00000001 (1) >
<2 x i32> %#2 = < #xffffffd6 (4294967254, -42), #xfffffff9 (4294967289, -7) >
<2 x i32> %#3 = < #x00000001 (1), #x00000001 (1) >
<2 x i32> %#4 = < #xffffffd6 (4294967254, -42), #xfffffff9 (4294967289, -7) >

Target:
Source value: < #xffffffd6 (4294967254, -42), #xfffffff9 (4294967289, -7) >
Target value: < #xffffffd6 (4294967254, -42), #x00000007 (7) >

Summary:
  0 correct transformations
  1 incorrect transformations
  0 failed-to-prove transformations
  0 Alive2 errors
```

### PR DESCRIPTION
**Pull request by dtcxzyw**
Previously we fold `div/rem X, C` into `poison` if any element of the constant divisor `C` is zero or undef. However, it is incorrect when threading udiv over an vector select: https://alive2.llvm.org/ce/z/3Ninx5
```
define <2 x i32> @vec_select_udiv_poison(<2 x i1> %x) {
  %sel = select <2 x i1> %x, <2 x i32> <i32 -1, i32 -1>, <2 x i32> <i32 0, i32 1>
  %div = udiv <2 x i32> <i32 42, i32 -7>, %sel
  ret <2 x i32> %div
}
```
In this case, `threadBinOpOverSelect` folds `udiv <i32 42, i32 -7>, <i32 -1, i32 -1>` and `udiv  <i32 42, i32 -7>, <i32 0, i32 1>` into `zeroinitializer` and `poison`, respectively. One solution is to introduce a new flag indicating that we are threading over a vector select. But it requires to modify both `InstSimplify` and `ConstantFold`.

However, this optimization doesn't provide benefits to real-world programs:
https://dtcxzyw.github.io/llvm-opt-benchmark/coverage/data/zyw/opt-ci/actions-runner/_work/llvm-opt-benchmark/llvm-opt-benchmark/llvm/llvm-project/llvm/lib/IR/ConstantFold.cpp.html#L908
https://dtcxzyw.github.io/llvm-opt-benchmark/coverage/data/zyw/opt-ci/actions-runner/_work/llvm-opt-benchmark/llvm-opt-benchmark/llvm/llvm-project/llvm/lib/Analysis/InstructionSimplify.cpp.html#L1107

This patch moves the fold into InstCombine to avoid breaking numerous existing tests. If necessary, I can split the NFC changes into another PR.

Fixes #114191 and #113866 (only poison-safety issue).

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -1158,29 +1158,39 @@ static Value *foldIDivShl(BinaryOperator &I, InstCombiner::BuilderTy &Builder) {
...
 
+  // If any element of a constant divisor fixed width vector is zero or undef
+  // the behavior is undefined and we can fold the whole op to poison.
+  if (auto *Op1C = dyn_cast<Constant>(Op1)) {
+    if (auto *VTy = dyn_cast<FixedVectorType>(Ty)) {
+      unsigned NumElts = VTy->getNumElements();
+      for (unsigned i = 0; i != NumElts; ++i) {
+        Constant *Elt = Op1C->getAggregateElement(i);
+        if (Elt && (Elt->isNullValue() || isa<UndefValue>(Elt)))
+          return replaceInstUsesWith(I, PoisonValue::get(Ty));
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1175]: We can convert it into an immediate UB here. But it's not worth the effort.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1175]: Why does moving this to InstCombine prevent us from converting the `udiv <i32 42, i32 -7> <i32 0, i32 1>` part to `poison`?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1175]: It doesn't block `udiv <i32 42, i32 -7> <i32 0, i32 1> -> poison` if the udiv is indeed an instruction. It just blocks this transform when threading udiv over selects.

**Comment by goldsteinn** [comment]
> If necessary, I can split the NFC changes into another PR.

Can you make it a separate commit in this PR?

File: llvm/lib/IR/ConstantFold.cpp [hunk 2]
```diff
@@ -902,11 +902,6 @@ Constant *llvm::ConstantFoldBinaryInstruction(unsigned Opcode, Constant *C1,
         Constant *ExtractIdx = ConstantInt::get(Ty, i);
         Constant *LHS = ConstantExpr::getExtractElement(C1, ExtractIdx);
         Constant *RHS = ConstantExpr::getExtractElement(C2, ExtractIdx);
-
-        // If any element of a divisor vector is zero, the whole op is poison.
-        if (Instruction::isIntDivRem(Opcode) && RHS->isNullValue())
-          return PoisonValue::get(VTy);
```
Comment (goldsteinn) [inline llvm/lib/IR/ConstantFold.cpp:908]: Could you just do `Res = PoisonValue::get(RHS->getType())` and make a poison element?

Comment (dtcxzyw) [reply llvm/lib/IR/ConstantFold.cpp:908]: Isn't it handled by the scalar path? See test `vec_select_udiv_poison`. `udiv <i32 42, i32 -7>, <i32 0, i32 1>` is folded into `<i32 poison, i32 -7>`.

**Review by nikic — APPROVED** [review]
LGTM

After some consideration, while it is technically the threadBinOpOverSelect fold that is incorrect here, being able to treat div/rem as lane-wise operations for simplification purposes does seem like a valuable property to have.

I'd note that in InstCombine, instead of generating a poison return, we could produce a non-terminator unreachable instead. Though it's kind of weird to do that only for the vector case. It might make sense to move all the division by zero folding to InstCombine... I could see this going either way in terms of overall simplification power.

**Pull request merged**